### ISSUE TITLE
Is the first call value for SoundMode.ringerModestatus unconditionally unknown?

### ISSUE THREAD

**Issue by ilsong963** [TryingOutSomething/sound_mode#12]
Thank you for making a good library and please understand that I use a translator.

I try to make a button to know the sound state of the device, but when I call it, unknown always returns, so I have to press it more than once.

I think it's a strange way to let a user click more than once.

In the example code, the initState calls first and the button calls again at onpressed.
I wonder if the reason for writing this is because the unknown value is always returned for the first call.

Additionally, I am going to use stateless widget, so it is also considered to change to stateful widget to have initState.

Is there a way to get a sound state value that is not unknown at once or is there an appropriate solution for my situation?

**Comment by TryingOutSomething** [comment]
Hi @ilsong963, thank you for checking out the plugin!

Can I check if you're testing the example on Android or iOS? Because I'm not able to reproduce the error on Android.

Edit: should be plugin and not library

**Comment by ilsong963** [comment]
The example code used in the movie.

In the code you uploaded, I commented out the _getCurrentSoundMode() in initState and deleted the buttons and functions that are not needed to implement the issue. And I added the touchCount code for button click confirmation.

It's almost the same as the example code.

There was no problem with Android, and only iOS was checked. The version of ios tested is 14.8.1.

---
There is the same problem even if I check it after updating to the latest version 15.5.

```
import 'dart:async';

import 'package:flutter/material.dart';
import 'package:sound_mode/sound_mode.dart';
import 'package:sound_mode/utils/ringer_mode_statuses.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatefulWidget {
  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  RingerModeStatus _soundMode = RingerModeStatus.unknown;
  String? _permissionStatus;
  int touchCount = 0;

  @override
  void initState() {
    super.initState();
    // _getCurrentSoundMode();
  }

  Future<void> _getCurrentSoundMode() async {
    RingerModeStatus ringerStatus = RingerModeStatus.unknown;
    try {
      ringerStatus = await SoundMode.ringerModeStatus;
    } catch (err) {
      ringerStatus = RingerModeStatus.unknown;
    }
    Future.delayed(const Duration(seconds: 1), () async {
      setState(() {
        _soundMode = ringerStatus;
        touchCount++;
      });
    });
  }

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        appBar: AppBar(
          title: const Text('Plugin example app'),
        ),
        body: Center(
          child: Column(
            mainAxisAlignment: MainAxisAlignment.center,
            children: <Widget>[
              Text('Running on: $_soundMode'),
              Text('Permission status: $_permissionStatus'),
              Text('touch count: $touchCount'),
              const SizedBox(height: 20),
              ElevatedButton(
                onPressed: () => _getCurrentSoundMode(),
                child: const Text('Get current sound mode'),
              ),
            ],
          ),
        ),
      ),
    );
  }
}
```


https://user-images.githubusercontent.com/50066666/170957805-425fa0e6-0c15-44b2-8589-247d6d534aac.MP4

**Comment by TryingOutSomething** [comment]
Hi @ilsong963, I apologise for the issue you faced while trying out the plugin. Can I check if you are running the example code on a simulator or an actual device? Because the plugin will not work while running on a simulator.

Unfortunately, I am unable to debug the iOS side of the plugin because I do not own a mac device. I deeply apologise for my inability to resolve this issue.

**Comment by ilsong963** [comment]
That video was taken with a actual device.

**Comment by TryingOutSomething** [comment]
Unfortunately, I cannot fix this issue because I do not own a mac device.

I tried to trace past pull requests and Googled the issue for iOS. It turns out that there is no native API to detect the sound mode on an iOS device, and the plugin is using this [third-party library](https://github.com/akramhussein/Mute) for the iOS method channel implementation. It seems that someone reported a similar issue, https://github.com/akramhussein/Mute/issues/15,a while back, but there was no follow up because the owner had no time to debug it.

A possible workaround would be to call `await SoundMode.ringerModeStatus` twice in the initState method.

Once again, I deeply apologise for the issue you faced while trying the plugin and my inability to fix this issue. Thank you for checking out the plugin.

**Comment by ilsong963** [comment]
I read akramhussein/Mute#15 and I don't think it's going to be a solution...

When I searched for it, there are VSSilentSwitch, MuteChecker, SoundSwitch, and MuteDetector besides Mute.

I don't have the ability to analyze swift code or make a flutter plugin.

If you have any thoughts of replacing it with another library, I will study and try. This is a constant problem for me, too.

**Comment by ilsong963** [comment]
Looking at [akramhussin/Mute#15](https://github.com/akramhussein/Mute/issues/15), this library problem is unlikely to be solved...

When I searched for it, there are [MuteChecker](https://github.com/sumitmundra/MuteChecker), [SoundSwitch](https://github.com/moshegottlieb/SoundSwitch), and [MuteDetector](https://github.com/fopina/MuteDetector) besides Mute.

I don't have the ability to analyze swift code or make a flutter plugin.

If you have any thoughts of replacing it with another library, I will study and try. This is a constant problem for me, too.

**Comment by TryingOutSomething** [comment]
Hi @ilsong963, thank you for suggesting alternatives to the Mute library for iOS.

I am okay with replacing the current Mute library with [SoundSwitch](https://github.com/moshegottlieb/SoundSwitch) because it is the 'predecessor' of the Mute library. However, `sound_mode` is using Swift language for the iOS side. I do not know how to change the plugin's iOS language to Objective-C or if we can import the Objective-C library to the current Swift implementation.

**Comment by fullflash** [comment]
yes for real device this plugin unable to detect sound mode it is always "unknown"
